src/mcts/node.h

```diff
@@ -570,6 +591,98 @@ class Edge_Iterator : public EdgeAndNode {
   uint16_t total_count_ = 0;
 };
 
+// TODO(crem) Replace this with less hacky iterator once we support C++17.
+// This class has multiple hypostases within one class:
+// * Range (begin() and end() functions)
+// * Iterator (operator++() and operator*())
+// It's more customary to have those as two classes, but
+// creating zoo of classes and copying them around while iterating seems
+// excessive.
+//
+// All functions are not thread safe (must be externally synchronized).
+template <bool is_const>
+class VisitedNode_Iterator {
+ public:
+  // Creates "end()" iterator.
+  VisitedNode_Iterator() {}
+
+  // Creates "begin()" iterator. Also happens to be a range constructor.
+  // child_ptr will be nullptr if parent_node is solid children.
+  VisitedNode_Iterator(const Node& parent_node, Node* child_ptr)
+      : node_ptr_(child_ptr),
+        total_count_(parent_node.num_edges_),
+        solid_(parent_node.solid_children_) {
+    if (node_ptr_ != nullptr && node_ptr_->GetN() == 0) {
+      operator++();
+    }
+  }
+  // These are technically wrong, but are usable to compare with end().
+  bool operator==(const VisitedNode_Iterator<is_const>& other) const {
+    return node_ptr_ == other.node_ptr_;
+  }
+  bool operator!=(const VisitedNode_Iterator<is_const>& other) const {
+    return node_ptr_ != other.node_ptr_;
+  }
+
+  // Function to support range interface.
+  VisitedNode_Iterator<is_const> begin() { return *this; }
+  VisitedNode_Iterator<is_const> end() { return {}; }
+
+  // Functions to support iterator interface.
+  // Equality comparison operators are inherited from EdgeAndNode.
+  void operator++() {
+    if (solid_) {
+      while (++current_idx_ != total_count_ &&
+             node_ptr_[current_idx_].GetN() == 0) {
+        if (node_ptr_[current_idx_].GetNInFlight() == 0) {
+          // Once there is not even n in flight, we can skip to the end. This is
+          // due to policy being in sorted order meaning that additional n in
+          // flight are always selected from the front of the section with no n
+          // in flight or visited.
+          current_idx_ = total_count_;
+          break;
+        }
+      }
+      if (current_idx_ == total_count_) {
+        node_ptr_ = nullptr;
+      }
+    } else {
+      do {
+        node_ptr_ = node_ptr_->sibling_.get();
+        // If n started is 0, can jump direct to end due to sorted policy
+        // ensuring that each time a new edge becomes best for the first time,
+        // it is always the first of the section at the end that has NStarted of
+        // 0.
+        if (node_ptr_ != nullptr && node_ptr_->GetN() == 0 &&
+            node_ptr_->GetNInFlight() == 0) {
+          node_ptr_ = nullptr;
+          break;
+        }
+      } while (node_ptr_ != nullptr && node_ptr_->GetN() == 0);
+    }
+  }
+  Node* operator*() {
+    if (solid_)
+      return &(node_ptr_[current_idx_]);
```

Comment:
Braces around the statement as it's in a separate line (same for `else`)

Reply:
Hmm yes not sure how this one snuck in. (They usually happen when one liners get autoformatted by clang-format, but that doesn't seem to be the case here.)